### DISCUSSION TITLE
Computing the film plane distance in perspective camera ?

### DISCUSSION THREAD

**Discussion by xacond00** [mitsuba-renderer/mitsuba3#1347] [Q&A] [answered]
I was looking into implementation of perspective camera, and noticed "importance" function doesn't scale with film plane distance squared at all.
Every other implementation from scientific papers defines the pdf/importance as:
`(film_distance)^2 / (A * cos(theta)^3)`
but yours doesn't ?
Specifically I mean these paper and their related sources:
https://bfraboni.github.io/mvpt19/index.html
https://bfraboni.github.io/eg2022/index.html
Could you explain why there might be such a discrepancy ?

Only thing that comes to mind, is that the film distance in Mitsuba 3 is always normalized to one ? But I can't really understand why, and how ? The "film" dimensions in that case cannot be physically correct ? Ie. when I look at bounds of generated samples in camera space, they fall into around (-1,1) at `z_near` distance, even though shouldn't they physically fit into a 35mm film ?  
But then again, the papers also define the film plane area and distance related to resolution and not really any physical dimensions...

Should I just consider the `m_normalization` to correspond to the whole `d^2 / A` term then ? 
Or how should I go around computing the film distance in Mitsuba ? I've tried: 
`m_film_dist = 0.5f / dr::tan(dr::deg_to_rad(m_x_fov * 0.5f));`,
however it's missing the scale in the nominator, and I have no idea how it should be related to the computed film area:
```c++
Point3f pmin(m_sample_to_camera * Point3f(0.f, 0.f, 0.f)), pmax(m_sample_to_camera * Point3f(1.f, 1.f, 0.f));  
m_image_rect.reset();
m_image_rect.expand(Point2f(pmin.x(), pmin.y()) / pmin.z());
m_image_rect.expand(Point2f(pmax.x(), pmax.y()) / pmax.z());  
m_normalization  = 1.f / m_image_rect.volume();  
```

Any help is much appreciated. And sorry for bothering again, I really cannot find any sources, which might explain this topic a little bit better...

**Comment by njroussel** [comment, marked as answer]
Hi @xacond00 

I'm unfamiliar with those works, so take the following with a pinch of salt - there might be some explanations that are specific to whatever setup they are doing/using in their work.

Fundamentally, where the film physically lies on the camera's z axis is irrelevant (assuming you're scaling it to keep the correct fov). As you've pointed out, we chose `d=1` in `PerspectiveCamera::importance()`, so the measure should now be `1 / (A * cos(theta)^3)`. The `1/A` term of that measure is indeed computed in `m_normalization`. Semantically, it should be equivalent to inverse area occupied by the film in camera space. So to compute it, we must figure out where the top-left corner and bottom right corners of the film are in camera space. The trick is to to use their respective sample space equivalent, i.e `(0, 0, 0)` and `(1,1,1)` and map them to camera space using `m_sample_to_camera`. From there we can compute the area using the last 4 lines of the snippet you copied.
I guess that the important bit is that `m_sample_to_camera` is coherent with the seemingly arbitrary choice of `d=1`. If you have a look at `Transform::perspective()` it is explicitly built such that points are projected onto a plane at `z=1` in camera space, hence `d=1`.
I hope this helped.

**Reply by xacond00** [reply]
> where the film physically lies on the camera's z axis is irrelevant (assuming you're scaling it to keep the correct fov

I think this is what I wanted to hear. The film distance stays constant, while the whole film scales instead to account for fov.
> So to compute it, we must figure out where the top-left corner and bottom right corners of the film are in camera space.

Yes, this is the one part I understood. However I got really confused by the "other sources" using the actual film resolution to compute the area, instead of some "projected" plane area, which was indeed quite weird.

I think the `film_distance^2` measure was used there to account for the film area being scaled with the film resolution, thus my main concern was that my `film distance`, has the same relative scale to the film plane area. Which it apparently does, when the film plane distance is indeed normalized to 1 and film area scales according to the fov, instead of the other way around.
 
One other thing, in thin lens camera, the film distance should be just equal to `m_focus_distance`, right ? As I see, in the sample_ray method the point is scaled to fit onto a "focal plane", instead of a film plane:
```c++
// Sampled position on the focal plane
Point3f focus_p = near_p * (m_focus_distance / near_p.z());
```

I'm forever grateful, and will contribute to the project, once my research is done.

**Reply by njroussel** [reply]
> One other thing, in thin lens camera, the film distance should be just equal to m_focus_distance, right ?

Yes, I believe so.
